handel/src/aggregation.rs

```diff
@@ -162,220 +185,202 @@ where
     fn automatic_update(&mut self) {
-        // Skip level 0 since it only contains this node
-        for level_id in 1..self.levels.len() {
-            let (receive_complete, next_peers) = {
-                let level = self.levels.get(level_id).unwrap();
-                (
-                    level.receive_complete(),
-                    level.select_next_peers(self.config.peer_count),
-                )
-            };
-
-            // Get the current best aggregate from store (no clone() needed as that already happens within the store)
-            // freeing the lock as soon as possible for the aggregation to continue.
-            let aggregate = {
-                let store = self.protocol.store();
-                let store = store.read();
-                store.combined(level_id - 1)
-            };
+        let store = self.protocol.store();
+        let store = store.read();
 
-            // For an existing aggregate for this level send it around to the respective peers.
-            if let Some(aggregate) = aggregate {
-                self.send_update(aggregate, level_id, !receive_complete, next_peers);
+        // Skip level 0 since it only contains this node.
+        for level_id in 1..self.levels.len() {
+            // Skip levels that aren't started yet.
+            let level = &self.levels[level_id];
+            if !level.is_started() {
+                break;
             }
+
+            // Send the aggregate to the next peers on this level.
+            let next_peers = level.select_next_peers(self.config.peer_count);
+            self.send_update(level_id, next_peers, &store);
         }
     }
 
-    /// activate the next level which needs to be activated.
+    /// Activate the next level that has not started yet.
     fn activate_next_level(&mut self) {
-        // the next level which needs activating on timeout.
-        let level = self.next_level_timeout;
-        // make sure such level exists
-        if level < self.levels.len() {
-            trace!(
-                id = ?self.protocol.identify(),
-                ?level,
-                "Timeout at level",
-            );
-
-            // next time the timeout triggers the next level needs activating
-            self.next_level_timeout += 1;
-
-            let store_rw = self.protocol.store();
-            let store = store_rw.read();
+        // Find the first level that has not been started yet.
+        let to_start = self.levels.iter().position(|level| !level.is_started());
 
-            // finally start the level.
-            self.start_level(level, &store);
-        }
-    }
+        // If there are no levels to start, we're done.
+        let Some(to_start) = to_start else {
+            return;
+        };
 
-    fn into_inner(self) -> (LevelUpdateStream<P, TId>, LevelUpdateSender<N>) {
-        (self.pending_contributions.into_stream(), self.sender)
+        // Start the level.
```

Comment:
If I understand correctly this used to start a level every set duration `d`. Meaning that if every level but the last was started immediately you would have to wait `(level_count - 1) * d`. Now it should start after just one times `d` elapses. Note that `Aggregation::start_level` does also reset this interval.

Basically before the change every `d` a level would attempted to be started, no matter the current circumstances. Now a new level will be started after `d` has elapsed since the previous level was started.